Доработана функциональность по техническому заданию 11 спринта

asdbl opened the pull request.

asdbl committed:
Добавить возможность ставить и убирать лайки фильмам.
Добавить возможность выводить популярные фильмы.
Добавить возможность добавлять и удалять друзей.
Добавить возможность выводить список общих друзей у пользователя.

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,29 @@
+
+@RestControllerAdvice("ru.yandex.practicum.filmorate")
+@Slf4j
+public class ErrorHandler {
```
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем `Exception`, а возвращать статус 500 INTERNAL_SERVER_ERROR
Так же следует добавить обработчик для исключения, которое выбрасывается при валидации данных с помощью аннотаций - `MethodArgumentNotValidException`, возвращать будет статус 400 BAD_REQUEST, так как ошибка связана с некорректными данными от пользователя)

TyutterinYakov replied:
Так же можно было бы не создавать для каждого исключения, обрабатываемого как `BAD_REQUEST`, отдельные хендлеры, а собрать их в одну группу в `@ExceptionHandler({MethodArgumentNotValidException.class, ValidationException.class...})`
А принимаемым аргументом метода-обработчика будет общий класс-родитель - `Exception`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,53 +1,58 @@
-    private Long id = 0L;
 
-    @PostMapping()
+    private final InMemoryFilmStorage inMemoryFilmStorage;
```
Уровень контроллер не должен напрямую работать с уровнем доступа к данным) Он должен это делать через прослойку в виде сервиса)
Также, когда будешь делать внедрение в сервисе:
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации) Например, при переходе на хранение в БД потребуется в разы меньше изменений, либо же при смене библиотеки для работы с БД, так как сервис не знает, какая из реализаций вообще используется и как осуществляется хранение

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,56 +1,63 @@
 public class UserController {
-    private final Map<Long, User> users = new HashMap<>();
-    private Long id = 0L;
+    private final InMemoryUserStorage inMemoryUserStorage;
```
Уровень контроллер не должен напрямую работать с уровнем доступа к данным) Он должен это делать через прослойку в виде сервиса)
Также, когда будешь делать внедрение в сервисе:
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации) Например, при переходе на хранение в БД потребуется в разы меньше изменений, либо же при смене библиотеки для работы с БД, так как сервис не знает, какая из реализаций вообще используется и как осуществляется хранение

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/exceptions/ErrorResponse.java:
```diff
@@ -0,0 +1,10 @@
+@AllArgsConstructor
+public class ErrorResponse {
+    @Getter
+    private String error;
```
Можно было бы сделать `final`, так как инициализация прооисходи через параметры конструктора и более значение изменяться не должно)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,53 +1,58 @@
+    }
+
+    @GetMapping("/popular")
+    public List<Film> getPopularFilms(@RequestParam(defaultValue = "10") String limit) {
```
Так как количество не может быть отрицательным, можно было бы реализовать валидацию с помощью аннотации) Для того, чтобы на вход поступало только корректное значение, следует перед аргументом метода поставить аннотацию `@Positive`, а над классом контроллера навесить аннотацию `@Validated`, таким образом перед передачей значения в метод, оно будет проверяться на соответствие, если оно не будет удовлетворять условию, будет выброшена ошибка валидации и обработана, а статус вернется красивым(400 BAD_REQUEST), а для красивого ответа от сервера нужно использовать еще один обработчик
```
@ExceptionHandler
public ResponseEntity<String> ....(ConstraintViolationException ex){
    ......
     return new ResponseEntity<>(ex.getMessage(), HttpStatus.BAD_REQUEST);
}
```

TyutterinYakov replied:
Можно было бы принимать значение сразу же в виде `Integer`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,71 @@
+@RequiredArgsConstructor
+@Slf4j
+public class UserService {
+    private final InMemoryUserStorage userStorage;
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации) Например, при переходе на хранение в БД потребуется в разы меньше изменений, либо же при смене библиотеки для работы с БД, так как сервис не знает, какая из реализаций вообще используется и как осуществляется хранение

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,57 @@
+@Slf4j
+public class FilmService {
+
+    private final InMemoryFilmStorage filmStorage;
```
Внедрение следует произвести через интерфейс хранилища, чтобы не привязываться к конкретной его реализации, а работать лишь с предоставляемым API, не задумываясь о реализации) Например, при переходе на хранение в БД потребуется в разы меньше изменений, либо же при смене библиотеки для работы с БД, так как сервис не знает, какая из реализаций вообще используется и как осуществляется хранение

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -3,26 +3,26 @@
     private LocalDate releaseDate;
     @Positive(message = "Длительность фильм не может быть отрицательной")
     private int duration;
+    private Set<Long> likes;
```
Можно было бы избавиться от всех проверок с инициализацией и произвести инициализацию здесь)
А над самим полем навесить `@JsonIgnore`, чтобы оно не учитывалось при сериализации-десериализации)

(предложение) Но лучше было бы представить это значение в виде поля на уровне хранилища) Так как модель не должна отвечать за хранение данных, она лишь их представляет)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -18,4 +19,5 @@ public class User {
     @NotNull
     @PastOrPresent(message = "День рождения не может быть в будущем")
     private LocalDate birthday;
+    private Set<Long> friends;
```
Можно было бы избавиться от всех проверок с инициализацией и произвести инициализацию здесь)
А над самим полем навесить `@JsonIgnore`, чтобы оно не учитывалось при сериализации-десериализации)

(предложение) Но лучше было бы представить это значение в виде поля на уровне хранилища) Так как модель не должна отвечать за хранение данных, она лишь их представляет)

Например, будет мапа в репозитории пользователей в виде 
`Map<Long, Set<Long>> friendsByUserId`

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,59 @@
+    private Long id = 0L;
+
+    @Override
+    public Film addFilm(@RequestBody @Valid Film film) {
```
Действительно ли эти аннотации требуются в этом классе?) Они нужны лишь на уровне контроллера

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Override
+    public User getUser(Long id) {
```
(Предложение) Хорошо было бы использовать в методах получения по идентификатору тип `Optional` https://habr.com/ru/post/346782/ Гарантирую, ты не пожалеешь, а заодно в будущем тебе будет проще, так как он активно используется при работе со Spring Data JPA

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public Film getFilm(long id) {
```
(Предложение) Хорошо было бы использовать в методах получения по идентификатору тип `Optional` https://habr.com/ru/post/346782/ Гарантирую, ты не пожалеешь, а заодно в будущем тебе будет проще, так как он активно используется при работе со Spring Data JPA

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private void userIdValidation(Long userId, Long friendId) {
+        if (userStorage.getUser(userId) == null) {
```
Если будешь использовать тип Optional, то здесь как раз будет проверка, можно будет использовать простую лямбду `.orElseThrow(() -> new NotFoundExcetion(...));` То есть, если объект не будет найден, будет выброшено исключение, а если он есть, мы просто его получим, как будто написали .`get()`
Это для метода получения по идентификатору, к которому, кстати, можно потом обращаться, чтобы не дублировать строку с выбросом исключения)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private void filmValidation(long userId, long filmId) {
+        if (userStorage.getUser(userId) == null) {
```
Если будешь использовать тип Optional, то здесь как раз будет проверка, можно будет использовать простую лямбду `.orElseThrow(() -> new NotFoundExcetion(...));` То есть, если объект не будет найден, будет выброшено исключение, а если он есть, мы просто его получим, как будто написали .`get()`
Это для метода получения по идентификатору, к которому, кстати, можно потом обращаться, чтобы не дублировать строку с выбросом исключения)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,57 @@
+        log.info("Getting popular films from limit {}", intLimit);
+        return filmStorage.getAllFilms()
+                .stream()
+                .filter(f -> f.getLikes() != null)
```
Проверка на `null` более не потрбуется

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего вечера! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌃

asdbl committed:
Хранение лайков фильмов перенесено на уровень хранилища.
Хранение друзей пользователей перенесено на уровень хранилища.
Взаимодействие с хранилищем перенесено на уровень сервиса.

asdbl committed:
Исправить checkstyle.

asdbl committed:
Исправить checkstyle.

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,37 @@
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public ErrorResponse handleConstraintViolation(ConstraintViolationException e) {
+        return new ErrorResponse(e.getMessage());
```
А здесь добавить логгирование?)
Также, почему бы не описать этот обработчик в том, что выше?)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,37 @@
+
+@RestControllerAdvice("ru.yandex.practicum.filmorate")
+@Slf4j
+public class ErrorHandler {
```
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем `Exception`, а возвращать статус 500 INTERNAL_SERVER_ERROR
Так же следует добавить обработчик для исключения, которое выбрасывается при валидации данных с помощью аннотаций - `MethodArgumentNotValidException`, возвращать будет статус 400 BAD_REQUEST, так как ошибка связана с некорректными данными от пользователя)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -3,26 +3,25 @@
     private LocalDate releaseDate;
     @Positive(message = "Длительность фильм не может быть отрицательной")
     private int duration;
+    private int like;
```
Это поле не требуется и его можно исключить) Сортировка, в твоем случае, происходит по содержимому мапы с лайками

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,111 @@
+                .toList();
+    }
+
+    private static void userValidation(User user) {
```
Следует вынести эту логику на уровень сервиса, так как это часть бизнес-правил)

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    @Override
+    public User getUser(Long id) {
```
Метод репозитория должен возвращать упакованное значение, а уже на уровне сервиса мы будем производить распаковку так, как нам требуется) Иначе мы не решаем проблему с тем, что код завязан лишь на одно возможное поведение

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Override
+    public Film getFilm(long id) {
```
Метод репозитория должен возвращать упакованное значение, а уже на уровне сервиса мы будем производить распаковку так, как нам требуется) Иначе мы не решаем проблему с тем, что код завязан лишь на одно возможное поведение

TyutterinYakov reviewed (commented):
Добрый день! Отметил оставшиеся недочеты) Хорошего дня! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌅

asdbl committed:
Метод получения из репозитория теперь возвращает Optional.
Добавить хендлер для необработанных исключений.

TyutterinYakov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    public Film getFilm(long filmId) {
+        filmValidation(filmId);
```
Вызов метода не имеет смысла, так как ниже мы и так выбросим исключение при отсутствующем значении

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного вечера!🌃🎇

asdbl committed:
Удалить вызов метода filmValidation в методе getFilm в классе FilmService
